### ISSUE TITLE
multiple build steps, parallel option

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#1356]
I'm proposing the ability to have multiple sub-sections in the `build` section of the yaml like this:

```
build:
  golang:
    image: golang
    commands:
      - go get
      - go build
  node:
    image: node
    commands:
      - npm install
      - npm run test
```

with the ability to specify that those sub-steps should run in parallel:

``` diff
build:
+  parallel: true
  golang:
    image: golang
    commands:
      - go get
      - go build
  node:
    image: node
    commands:
      - npm install
      - npm run test
```

we should provide the ability to limit parallelization. The above example will fanout all steps in parallel. The below example will limit to N parallel steps:

``` diff
build:
+  parallel: 4
```

note that as an added benefit, we could use the `where` clause to limit when steps are executed. For example, only execute certain builds steps for certain branches, or for certain events (ie not pull request):

``` diff
build:
  golang:
    image: golang
    commands:
      - go get
      - go build
  node:
    image: node
    commands:
      - npm install
      - npm run test
+  dist:
+    image: golang
+    commands:
+      - make dist
+      - make deb
+    when:
+      event: push
```

some more use cases for the above build steps and `when` clause might be:
- avoid running certain resource-intensive steps during pull requests
- only run a subset of tests for pull requests, run full test suite on merge
- execute deployment-related steps when not a pull request

---

until we have the ability to define multiple build sections, the recommended **workaround** is to execute any bash deployment steps inside your build section:

```
build:
  image: golang
  commands:
    - go build
    - go test
    - ./deploy.sh
```

note that you may not want to execute certain scripts for pull requests. If the build is a pull request you can look for this environment variable and value: `CI_EVENT=pull_request`

**Comment by Joshua-Anderson** [comment]
I like this idea, but how would output be displayed for parallel build steps? It could be really hard to read the logs if two process are printing to it at the same time.

**Comment by bradrydzewski** [comment]
The output would likely come over a single stream, and would need to be separated ... this is the big technical challenge here. Once separated it becomes a UX problem and I'm sure we can test out a bunch of different options to visualize the output.

First step is really to enable multiple build steps running serially.

**Comment by gbergere** [comment]
It would be create to allow us to have different job (build part) in different branch. I mean different build for dev, qa, staging, ...
So +1

**Comment by macb** [comment]
I think `parallel` being a maximum number of parallel steps might be helpful. Imagine we have 2 build steps with 4 matrices (if we assume matrices run in parallel as they seem planned to) that could be 8 things running at once. For something like `go test ./...` this may have a negative effect on the duration since each instance would be using all CPU's and there would be a lot of contention.

So `parallel: 4` would mean only 4 steps could run at a time, with 4 pending while the first 4 start, then a new one starting as an old one finishes until all 8 are completed.

Or maybe an additional configuration for the limit.

**Comment by bradrydzewski** [comment]
@macb agreed. edited the above spec to include parallel limits

**Comment by jakubsacha** [comment]
I'd love to have something like "stages". Brief of concept is something like this:

```
stages:
- prepare
- unit
- integration

prepare:
   golang:
      image: golang
      commands: ...
   nginx:
      image: php
      commands: ...

unit:
   golang:
      image: ...

```
Concept is that all jobs in one stage are executed in parallel, and stages are executed sequentially. It's way more flexible than just *parallel* switch.

What do you think guys?

**Comment by bradrydzewski** [comment]
@jakubsacha we could probably infinitely nest sections, allowing for some really complex execution trees. Drone reads the yaml file into an abstract syntax tree and will therefore be able to execute some pretty complex sequences with ease

```diff
build:
  compile:
    image: golang
    commands:
      - go get
      - go build
+  test:
+    parallel: true
+    unit:
+      image: golang
+      commands:
+        - go test
+    selenium:
+      image: golang
+      commands:
+        - ./selenium.sh
```

I definitely think named top-level stages are interesting. We would need the following capabilities for something like this to work:

* enforce ordering of `clone` and `cache` steps
* identify which steps are plugins vs build containers vs compose containers
* identify which steps are `publish` and `deploy` to prevent execution during pull requests
* identify which steps are **not** plugins so that we don't inject secrets that are intended for plugins

We also need to maintain compatibility with the existing yaml. I think this would be possible because `build` would be interpreted as just another stage in the process.

I think there is merit to the idea, but we would need to address the points above.

**Comment by crhym3** [comment]
As an alternative to `parallel`, may I suggest dependent steps, e.g.

```yaml
build:
  compile:
    image: golang
    commands:
      - go get
      - go build
  test:
    unit:
      deps: [build]
      image: golang
      commands:
        - go test
    selenium:
      deps: [build]
      image: golang
      commands:
        - ./selenium.sh
```

but then nesting might even become irrelevant; the above could be rewritten as:

```yaml
build:
  compile:
    image: golang
    commands:
      - go get
      - go build
  unit:
    deps: [build]
    image: golang
    commands:
      - go test
  selenium:
    deps: [build]
    image: golang
    commands:
      - ./selenium.sh
```

and `parallel` would be just a function of steps dependencies, although a `concurrent: n` is probably still useful to limit the max number of tasks executed in parallel.

**Comment by bradrydzewski** [comment]
@crhym3 I like the second option. First one to implement their proposal wins!

**Comment by bradrydzewski** [comment]
This pull request demonstrates how we might enable multi-build parsing in the yaml. It allows both single build (current format) and multi-build with a single layer of subsections: https://github.com/drone/drone-exec/pull/15/files

This doesn't implement filtering but that should be trivial to add. This also doesn't attempt to address concurrency since we don't have a way to separate the output in the build logs.

**Comment by cusspvz** [comment]
Think this is my first comment here, as so I would like to congrat all contributors for this project.

We've started testing Drone since it has a small footprint and fast as hell. Unfortunately, INMO, there are some things that could be enhanced, or we had totally misunderstood the [readme](http://readme.drone.io) docs.

The main one is related with this issue. I've been using different CIs over time and I really liked the way GitLab CI have their recently new configuration parameters schema.

@jakubsacha has already commented about `stages` and thats exactly how *GitLab CI* is based up.

They have shorted most of configuration keys so they could bring up dynamic build `stages`. Here's how it works:

```
# In case you don't define any, by default you got a `test` stage, which could be represented as this:
stages:
- test

# But neither all CD are equal, some of us want to test after build, others want to test before the build, so, its your choice the order you need:
stages:
- build
- test
# or
stages:
- test
- build
```

Then you have `jobs`, `jobs` are part of a `stage` and they assume (correctly) that all jobs must run in parallel for better performance (thats the relation with this issue), so I can have 3 test suites running at same time and only start building in case all tests were right.

```
stages:
- test
- build
- deploy

test_example_1:
  stage: test
  commands:
  - echo "Hello World"

test_example_2:
  stage: test
  commands:
  - echo "World Hello"

test_example_3:
  stage: test
  commands:
  - echo "still behind build, running in parallel with test jobs above"

build_nothing:
  stage: build
  commands:
  - echo "I'm only executed after succeeded tests"

gce_deployment:
  stage: deploy
  commands:
  - echo "there's nothing to deploy... but still executed after a succeeded build"
```

There was a question I was about to post on another issue, but since this is turning up into a proposal, how do you attach or relate a plugin into a certain configuration?

Supposing that you relate plugin by job's name, it is not possible to use the same plugin more than once, right? If so, thats a *con* for us and many since most of well scalable projects have their docker services on different images (usually built from the same repo).

Here's my proposal: On each job you could set which plugin you want to use, here's an npm example with an email and webhook to a buildbell (hehehe) on finish.

`use` could default to `shell`, which will have usual behavior without plugins.

```

stages:
- test
- publish
- finish

jshint:
  stage: test 
  use: shell # well, just for PoC
  commands:
  - npm run test-jshint

mocha:
  stage: test
  commands:
  - npm install --development
  - npm run test-mocha

npm:
  stage: publish
  use: npm
  repository: https://npm-repo.example.org
  username: user
  password: pass

warn_octocat:
  stage: finish
  use: email
  from: noreply@github.com
  host: smtp.mailgun.org
  username: octocat
  password: 12345
  recipients:
  - octocat@github.com

ring_bell:
  stage: finish
  use: webhook
  urls:
  - http://bell.development.example.net/
  header:
    X-Music: "StarWars"

```

@gbergere also mentioned different build steps for different environments. Thats also possible with the use of `when` (cannot remember how to setup, also don't know if it is compatible with regexp).

```

stages:
- test

some_job_that_would_run_on_all_triggers:
  stage: test
  commands: 
  - echo "olereliliiii"

test_needed_on_production:
  stage: test
  use: shell
  commands:
  - echo "hey cide, yiu're ginna ti priductiin, ginna test yiu..."
  - run forest; run
  when:
    tag:
      matches: /[0-9]+\.[0-9]+\.0$/ #match all semver minor releases

```

**Comment by bradrydzewski** [comment]
@cusspvz first let me say that I appreciate you taking the time to write up suggestions.

In preparation for a 1.0 release we will not be making any breaking changes to the existing yaml format. There are a large number of Drone installs and when we announced the 0.4 beta that included a stability promise so that developers could feel comfortable investing time in the platform.

I do think it is important, however, to understand how other platforms (like GitLab) are solving these issues and to adopt good ideas from other projects when possible. Also, these sort of discussions will be taken into account if / when we choose to break backward compatibility (likely Drone 2.0)

So with that being said, I've taken the above example and converted to the Drone yaml format. I don't see anything in the above format that we can't accomplish today:

```
build:
  jshint:
    image: node
    commands:
      - npm run test-jshint

  mocha:
    image: node
    commands:
      - npm install --development
      - npm run test-mocha

publish:
  npm:
    repository: https://npm-repo.example.org
    username: user
    password: pass

notify:
  email:
    from: noreply@github.com
    host: smtp.mailgun.org
    username: octocat
    password: 12345
    recipients:
    - octocat@github.com

  webhook:
    urls:
    - http://bell.development.example.net/
    header:
      X-Music: "StarWars"
```

> @gbergere also mentioned different build steps for different environments. Thats also possible with the use of when (cannot remember how to setup, also don't know if it is compatible with regexp).

This is also supported in the current yaml format. You can read more about conditional step logic at the below links:

* http://readme.drone.io/usage/build_test/#conditions:fb92aa3346185c57f15afda861d465a3
* http://readme.drone.io/usage/deploy/#conditions:0814af543608f4fd171b2e9cf43d9917
* http://readme.drone.io/usage/notify/#conditions:d1fea7f63b32cc4bab67588d2fde8aa7

> Supposing that you relate plugin by job's name, it is not possible to use the same plugin more than once, right? If so, thats a con for us and many since most of well scalable projects have their docker services on different images (usually built from the same repo).

This is incorrect. See the last example at the bottom of this page http://readme.drone.io/usage/deploy/

**Comment by tboerger** [comment]
Maybe we should mention that the drone yml parser is different than most yaml parsers because it allows multiple times the same key.

**Comment by bradrydzewski** [comment]
@cusspvz I meant to comment on this as well

> use could default to shell, which will have usual behavior without plugins.

This is something the existing yaml design does poorly. We have no way to execute plugin steps during the build process (only before or after). I agree it would be nice to have more flexibility here.

**Comment by crhym3** [comment]
For completeness, here's the example https://github.com/drone/drone/issues/1356#issuecomment-170650041 converted to a format described in https://github.com/drone/drone/issues/1356#issuecomment-165939047 (execution tree as opposed to flat sequence)

```yaml
build:
  test_example_1:
    commands:
      - echo "Hello World"

  test_example_2:
    commands:
    - echo "World Hello"

  test_example_3:
    commands:
    - echo "still behind build, running in parallel with test jobs above"

  build_nothing:
    deps: [test_example1, test_example_2, test_example_3]
    commands:
    - echo "I'm only executed after succeeded tests"

deploy:
  # some gce deployment plugin
  gce:
    foo: bar
```

**Comment by cusspvz** [comment]
Thanks for your faster replies!!1

> these sort of discussions will be taken into account if / when we choose to break backward compatibility (likely Drone 2.0)

Didn't knew about that, thanks for mentioning...

> Maybe we should mention that the drone yml parser is different than most yaml parsers because it allows multiple times the same key.

Definitely. 

>> (cannot remember how to setup, also don't know if it is compatible with regexp)

> This is also supported in the current yaml format.

Know that this is kind of off-topic on this issue, but would the regexp `^[0-9]+.[0-9].0$` work with `when: branch` ?

**Comment by chuyskywalker** [comment]
As some extra reading, the Jenkins "Buildflow" plugin has some really great support for flow options, parallel, steps, calling other jobs, etc. Could provide some inspiration for what's being discussed here. 

https://wiki.jenkins-ci.org/display/JENKINS/Build+Flow+Plugin

**Comment by bradrydzewski** [comment]
@crhym3 suggest we change the name from `deps` to `depends_on` to keep naming consistency with Docker compose, but otherwise I think this is the approach we should adopt. Are you thinking that everything runs in parallel unless otherwise specified? Or that steps only run in parallel when a `depends_on` is found in subsequent steps?

```
build:
  frontend:
    image: node
    commands: [ npm install, npm test ] 
  backend:
    image: golang
    commands: [ go build, go test ]
  notify:
    image: slack
    depends_on: [ frontend, backend ]
```

**Comment by crhym3** [comment]
Yes, the idea was that every step runs in parallel implicitly. For instance, you first create the execution tree based on `depends_on` (edges) and then begin running the steps starting with the root. In your example it would be:

```
              root
              /  \
      frontend    backend
             \     /
              \   /
             notify
```

Although `notify` is kind of a special case - you want to run it regardless. So, it probably needs a flag, something like `run: always`.

**Comment by bradrydzewski** [comment]
@crhym3 so one minor consideration is the next version of drone will remove publish, deploy and notification section and instead use a single block in the yaml for all steps. This allows us to mix plugin steps and build steps in whatever order makes sense for the project, and even run plugins for pull requests (I believe something you were interested in also)

This is an example of what the yaml currently looks like:
http://readme.drone.io/0.5/usage/overview/#overview

With these Yaml change I believe we'll want sequential behavior by default to avoid plugins from otherwise executing in parallel, before the build steps complete. Perhaps instead we can take the proposed yaml changes and make parallelism opt-in? What if only steps defined in depends_on were run in parallel?

**Comment by nafg** [comment]
@bradrydzewski I think that's counterintuitive: A dependency is a "break" from potential parallelism. Sequentially is implicit dependency (finish this first, then start that).

I think Gitlab CI has a nice compromise between flexibility and conciseness: http://doc.gitlab.com/ce/ci/yaml/README.html#stages

**Comment by bradrydzewski** [comment]
@nafg re-designing the yaml isn't an option so we need something that fits into the current architecture

**Comment by nafg** [comment]
@bradrydzewski not much redesigning required, I linked to the specific section would could easily be adopted. Basically:

1. Every pipeline has a set of _stages_. The default is `["build", "test", "deploy"]`.
2. Every step has a stage. The default is `"test"`.

So you could write
```yaml
pipeline:
  stages:
    - build
    - test
  step1:
    stage: build
    ...
  step2:
     stage: build
     ...
  step3:
     ...
  step4:
     ...
```

Steps 1 and 2 would run in parallel, then 3 and 4 would.

It's not as expressive as being able to specify dependencies, but it's probably adequate for normal build sizes.

**Comment by cusspvz** [comment]
@nafg I've already proposed a [GitLab alike implementation](https://github.com/drone/drone/issues/1356#issuecomment-170650041). [This](https://github.com/drone/drone/issues/1356#issuecomment-170666953) was @bradrydzewski 's response.
Please read the whole discussion. :)

**Comment by bradrydzewski** [comment]
First thing to keep in mind the Go yaml parser in pretty inflexible since we are dealing with a static type system. The pipeline is a slice of containers which means adding a steps section would mix `[]string` with `[]container` and would require custom parsing, which I'd like to avoid when possible.

Second thing keep in mind that the .drone.yml is a superset of the docker-compose file. The concept of stages, in my opinion, does not align well with docker-compose. This introduces a new construct not used by the services section ...

I really don't think we would need to define a `stages` array anyway. The list of stages (or groups or whatever) could be extracted from the yaml at runtime:

```yaml
pipeline:
  {step}:
    image: {image}
    group: {parallel group} # this could be called stage or whatever
```

Let's say we have this Yaml file:

```yaml
pipeline:
  clone:
    image: git
    depth: 1

  frontend:
    image: node
    group: build # denotes parallelism
    commands: npm install

  backend:
    image: golang
    group: build  # denotes parallelism
    command: [ go build, go test ]

  deploy:
    image: docker
    repo: drone/drone
```

This would get executed as:

```
          clone
         /     \
        /       \
  frontend     backend   [grouped by "build"]
         \      /
          \    /
          deploy
```

I think `group` is probably not the ideal attribute because it may be in use by existing plugins, so we probably shouldn't make it a reserved word. This is something I'll need to verify, though.

**Comment by nafg** [comment]
@bradrydzewski how does it decide that clone goes first and deploy goes last?

**Comment by bradrydzewski** [comment]
@nafg drone executes steps sequentially in the order they are defined in the yaml. This has always been the behavior and, to my knowledge, has worked well.

**Comment by nafg** [comment]
@bradrydzewski so what would happen if the sequence was [ungrouped, group=A, ungrouped, group=A]? And whichever answer you pick, can you argue that it's intuitive?
It seems to me that parallelism and order-of-definition sequential are in conflict.

**Comment by nafg** [comment]
How about borrowing another detail from gitlab: instead of there being such a thing as ungrouped, we say default group = test (or whatever you pick). Then, you can run the default group in parallel too.

At that point, you might want to change group from a property to a header. So your example would become,

```yaml
pipeline:
 step1:
  clone:
    image: git
    depth: 1

 build:
  frontend:
    image: node
    commands: npm install
  backend:
    image: golang
    command: [ go build, go test ]

 step3:
  deploy:
    image: docker
    repo: drone/drone
```

**Comment by bradrydzewski** [comment]
>  so what would happen if the sequence was [ungrouped, group=A, ungrouped, group=A]? And whichever answer you pick, can you argue that it's intuitive?

I'm not sure I follow. I would need to see an example

> How about borrowing another detail from gitlab: instead of there being such a thing as ungrouped, we say default group = test (or whatever you pick). Then, you can run the default group in parallel too.

This adds complexity to the yaml for all projects, regardless of whether or not they need parallelism. It also breaks structural compatibility with the docker-compose yaml format, which should be avoided if possible.

I think the below proposal meets some important constraints:

* it maintains compatibility with how drone works today (sequential steps)
* maintains compatibility with the docker compose structure and hierarchy
* adds a simple way to define parallelism if you need it
* doesn't add extra blocks for projects not using parallelism

```yaml
pipeline:
  clone:
    image: git
    depth: 1

  frontend:
    image: node
    group: build # denotes parallelism
    commands: npm install

  backend:
    image: golang
    group: build  # denotes parallelism
    command: [ go build, go test ]

  deploy:
    image: docker
    repo: drone/drone
```

If there are any major issues with the above design I want to identify them now. If there are alternate proposals that meet the constraints defined above I'm interested in them. The 0.5 yaml (aside from parallelism) is locked down and __will not__ undergo any major changes so that should be considered as well.

**Comment by tboerger** [comment]
I'm totally fine with the group attribute

**Comment by MorrisJobke** [comment]
> That being said, if there are any major practical limitations to the above design I want to identify them now.

Currently the working directory is shared: Then the frontend and the backend steps would access it in the same time. What happens when those tests are integration tests and configure the software. This means that the both steps would overwrite each others configuration. For parallel executed steps it would make sense to provide each step executor a separate copy of the working directory. And what happens to services? Are then the same services linked? Or are they duplicated?

Let me give you a practical example:

We are an open source project that has many different configurations and different types of tests:

- 4 unit test runs (mysql, mariadb, sqlite, postgres)
- unit tests on different versions of the language (but only with sqlite)
- integration tests for API (based on one of the above DBs)
- end-to-end tests (a client that is run against a service container that holds the checked out code - maybe possible by starting a container in a trusted env?!?)
- JS unit tests (no need for any service container)
- code linters
- run special tests only against a given service container (run the tests for SMB-external storage with the SMB container, tests for S3 against S3 container, ...)

All of them could be run in parallel totally fine. The only constraint here is, that neither the working directory nor services are shared, because there could be interferences then.

Would it make sense to allow to specify service containers per (parallel) build step?

Also for me the group attribute looks fine.

**Comment by bradrydzewski** [comment]
Feature is implemented and should be included in drone core within the next two weeks
https://github.com/cncd/pipec/blob/master/samples/sample_5/pipeline.yml

**Comment by bradrydzewski** [comment]
This is merged into master using the syntax proposed [above](https://github.com/drone/drone/issues/1356#issuecomment-218626772). I definitely agree we could probably improve the syntax, but for now I think it is important to let this bake and gather feedback
